server.go

```diff
@@ -0,0 +1,720 @@
+	if pattern == "" {
+		panic("COAP: invalid pattern " + pattern)
+	}
+	mux.m.Lock()
+	delete(mux.z, pattern)
+	mux.m.Unlock()
+}
+
+// ServeCOAP dispatches the request to the handler whose
+// pattern most closely matches the request message. If DefaultServeMux
+// is used the correct thing for DS queries is done: a possible parent
+// is sought.
+// If no handler is found a standard NotFound message is returned
+func (mux *ServeMux) ServeCOAP(w ResponseWriter, request Message) {
+	h, _ := mux.match(request.PathString())
+	if h == nil {
+		h = failedHandler()
+	}
+	h.ServeCOAP(w, request)
+}
+
+// Handle registers the handler with the given pattern
+// in the DefaultServeMux. The documentation for
+// ServeMux explains how patterns are matched.
+func Handle(pattern string, handler Handler) { DefaultServeMux.Handle(pattern, handler) }
+
+// HandleRemove deregisters the handle with the given pattern
+// in the DefaultServeMux.
+func HandleRemove(pattern string) { DefaultServeMux.HandleRemove(pattern) }
+
+// HandleFunc registers the handler function with the given pattern
+// in the DefaultServeMux.
+func HandleFunc(pattern string, handler func(ResponseWriter, Message)) {
+	DefaultServeMux.HandleFunc(pattern, handler)
+}
+
+// A Server defines parameters for running an COAP server.
+type Server struct {
+	// Address to listen on, ":COAP" if empty.
+	Addr string
+	// if "tcp" or "tcp-tls" (COAP over TLS) it will invoke a TCP listener, otherwise an UDP one
+	Net string
+	// TCP Listener to use, this is to aid in systemd's socket activation.
+	Listener net.Listener
+	// TLS connection configuration
+	TLSConfig *tls.Config
+	// UDP "Listener" to use, this is to aid in systemd's socket activation.
+	PacketConn net.PacketConn
+	// Handler to invoke, COAP.DefaultServeMux if nil.
+	Handler Handler
+	// Default buffer size to use to read incoming UDP messages. If not set
+	// it defaults to 1500 B.
+	UDPSize int
+	// The net.Conn.SetReadTimeout value for new connections, defaults to 2 * time.Second.
+	ReadTimeout time.Duration
+	// The net.Conn.SetWriteTimeout value for new connections, defaults to 2 * time.Second.
+	WriteTimeout time.Duration
+	// If NotifyStartedFunc is set it is called once the server has started listening.
+	NotifyStartedFunc func()
+
+	TCPReadBufferSize  int
+	TCPWriteBufferSize int
+
+	readerPool sync.Pool
+	writerPool sync.Pool
+	// UDP packet or TCP connection queue
+	queue chan *requestCtx
+	// Workers count
+	workersCount int32
+	// Shutdown handling
+	lock    sync.RWMutex
+	started bool
+}
+
+func (srv *Server) worker(w *requestCtx) {
+	srv.serve(w)
+
+	for {
+		count := atomic.LoadInt32(&srv.workersCount)
+		if count > maxWorkersCount {
+			return
+		}
+		if atomic.CompareAndSwapInt32(&srv.workersCount, count, count+1) {
+			break
+		}
+	}
+
+	defer atomic.AddInt32(&srv.workersCount, -1)
+
+	inUse := false
+	timeout := time.NewTimer(idleWorkerTimeout)
+	defer timeout.Stop()
+LOOP:
+	for {
```

Comment:
try to avoid this, maybe extract to separate func and return instead of break ANCH

Reply:
done